File: i3/.i3/scripts/cycle_workspace.js
var exec = require('child_process').exec;

exec('i3-msg -t get_workspaces', function (error, stdout, stderr) {
  var workspaces = JSON.parse(stdout),
      current = workspaces.filter(function (x) { return x.focused })[0],
      maybeNext = workspaces.filter(function (x) { return x.output === current.output });
      next = ( maybeNext.indexOf(current) + 1 === maybeNext.length) ? 0 : maybeNext.indexOf(current) + 1;

  exec('i3-msg workspace ' + maybeNext[next].num);
});

File: i3/.i3/scripts/cycle_workspace.py
import json
import subprocess


I3MSG = '/usr/bin/i3-msg'


def main():
    result = subprocess.Popen([I3MSG, '-t', 'get_workspaces'],
                              stdout=subprocess.PIPE,
                              stdin=subprocess.PIPE)
    result = result.communicate()[0]
    data = json.loads(result)

    focused = filter(lambda x: x['focused'] == True, data)[0]
    maybeNext = filter(lambda x: x['output'] == focused['output'], data)

    if maybeNext.index(focused) + 1 == len(maybeNext):
        nextWorkspace = 0
    else:
        nextWorkspace = maybeNext.index(focused) + 1

    subprocess.call([I3MSG, 'workspace', str(maybeNext[nextWorkspace]['num'])])


if __name__ == '__main__':
    main()

